api/api.go

```diff
@@ -0,0 +1,129 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+)
+
+// WeatherReponse is used to accept JSON structured
+// data.  Output is not very human friendly and is thus
+// a stage type for the Weather struct
+type WeatherResponse struct {
+	Weather []struct {
+		Main        string `json:"main"`
+		Description string `json:"description"`
+	} `json:"weather"`
+	Main struct {
+		Temp float64 `json:"temp"`
+	} `json:"main"`
+	City string `json:"name"`
+}
+
+// Weather is the human friendly struct that is populated
+// by the WeatherReponse struct
+type Weather struct {
+	Main        string
+	Description string
+	Temp        float64
+	City        string
+}
+
+// Client is used to assemble the necessary parts for a HTTP request
+// and includes the HTTPClient lib to set timeout
+type Client struct {
+	Base       string
+	Units      string
+	ApiKey     string
+	HTTPClient *http.Client
+}
+
+// Get takes a location and resturns a Weather struct and error.
+func (c Client) Get(location string) (Weather, error) {
+
+	//assemble request based on Client struct data
+	url := c.FormatURL(location)
+
+	//use Client since we override default HTTPClient settings -- timeout
+	resp, err := c.HTTPClient.Get(url)
+
+	if err != nil {
+		return Weather{}, fmt.Errorf("something went wrong.  Please try again later.  %v", err)
+	}
+	//close when done to prevent resource leaks
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return Weather{}, fmt.Errorf("unexpected status code.  %v", resp.StatusCode)
+	}
+
+	// pass io.Reader to ParseReponse
+	return ParseResponse(resp.Body)
+
+}
+
+// NewClient takes apiKey and tempunits populates the Client struct
+// with a base url, temperature units, apikey AND sets up the
+// HTTPClient with timeout settings since default timeout is too long.
+// Returns Client struct and error
+func NewClient(apiKey string, tempunits string) (Client, error) {
+
+	var c Client
+
+	if tempunits == "metric" || tempunits == "imperial" || tempunits == "standard" {
```

Comment:
And what if it's not one of those? It seems like we just silently ignore it in that case. I think it probably makes sense either not to validate this at all—just pass it straight to the weather API and hope the user knows what they're doing—or to validate it strictly, reporting an error if it's not one of the allowed units.

A nice way to do the validation, by the way, is to create some named constants in your package, like `weather.UnitsMetric`, `weather.UnitsImperial`, and so on. Then the user can see what's available, because it'll be in the generated documentation for your package, and the compiler itself will tell them if there's a typo or something. Have a look at the 'Behaviour' book for some examples of this.

Reply:
Ok, I took at look at the book and incorporated the constant/map/func pattern to validate